monitoring_service/blockchain.py

```diff
@@ -1,83 +1,31 @@
-        self.is_running.set()
-        self.event_handlers = {
-            EVENT_CHANNEL_CLOSE: [],
-            EVENT_CHANNEL_SETTLED: [],
-            EVENT_TRANSFER_UPDATED: []
-        }
-        self.web3 = web3
-        self.poll_interval = 5
-        self.event_filters = None
-
-    def make_filters(self):
-        ret = []
-        for event_name in self.event_handlers.keys():
-            abi = CONTRACT_MANAGER.get_event_abi('TokenNetwork', event_name)
-            assert abi is not None
-            ret.append(make_filter(self.web3, abi))
-        return ret
-
-    def _run(self):
-        while self.is_running.is_set():
-            try:
-                if self.event_filters is None:
-                    self.event_filters = self.make_filters()
-                self.poll_blockchain()
-            except requests.exceptions.ConnectionError as e:
-                endpoint = self.web3.providers[0].endpoint_uri
-                log.warning(
-                    'Ethereum node (%s) refused connection. Retrying in %d seconds.' %
-                    (endpoint, self.poll_interval)
-                )
-                self.event_filters = None
-                gevent.sleep(self.poll_interval)
-        self.uninstall_filters()
-
-    def uninstall_filters(self):
-        if self.event_filters is None:
-            return
-        [self.web3.eth.uninstallFilter(f.filter_id)
-         for f in self.event_filters]
-        self.event_filters = None
-
-    def register_handler(self, event, callback):
-        self.event_handlers[event].append(callback)
-
-    def poll_blockchain(self):
-        for f in self.event_filters:
-            events = f.get_new_entries()
-            [self.handle_event(ev) for ev in events]
-        gevent.sleep(self.poll_interval)
-
-    def stop(self):
-        self.is_running.clear()
-
-    def handle_event(self, event):
+class BlockchainMonitor(BlockchainListener):
+    def __init__(self, web3, **kwargs):
+        super().__init__(
+            web3,
+            CONTRACT_MANAGER,
+            'TokenNetwork',
+            poll_interval=1,
+            **kwargs
+        )
+
+    def add_confirmed_listener(self, event_name: str, callback: callable):
+        """ Add a callback to listen for confirmed events. """
+        return super().add_confirmed_listener(
+            event_name,
+            lambda event: self.handle_event(event, callback)
+        )
+
+    def handle_event(self, event, callback):
         tx = self.web3.eth.getTransaction(event['transactionHash'])
         abi = CONTRACT_MANAGER.get_contract_abi('TokenNetwork')
         s = decode_contract_call(abi, tx['data'])
         assert s is not None
-        handlers = self.event_handlers.get(event['event'], None)
-        log.info(event)
-        if handlers is None:
-            log.warning('unhandled event type: %s' % str(event))
-            return
-        [x(event, s) for x in handlers]
+        return callback(event, s)
```

Comment:
What is `s` used for? We could alternatively return this in the BlockchainListener as well (maybe optionally)

Reply:
It's decoded method parameters*, returned as a simple list. I am not sure whether to put it to the listener - it affects performance a little. But if you find use for it in PFS it can be shared somehow.

*  (though it's not a best variable name, I'll update that)

Reply:
Let's leave it here for now, and change it when it becomes necessary.